tractor/trionics/_supervisor.py

```diff
@@ -0,0 +1,256 @@
+
+        return self.result
+
+
+class TaskHandle(Struct):
+    task: Task
+    cs: CancelScope
+    exited: Event | None = None
+    _outcome: Outcome | None = None
+
+
+class ScopePerTaskNursery(Struct):
+    _n: Nursery
+    _scopes: dict[
+        Task,
+        tuple[CancelScope, Outcome]
+    ] = {}
+
+    scope_manager: ContextManager | None = None
+
+    async def start_soon(
+        self,
+        async_fn,
+        *args,
+
+        name=None,
+        scope_manager: ContextManager | None = None,
+
+    ) -> tuple[CancelScope, Task]:
+
+        # NOTE: internals of a nursery don't let you know what
+        # the most recently spawned task is by order.. so we'd
+        # have to either change that or do set ops.
+        # pre_start_tasks: set[Task] = n._children.copy()
+        # new_tasks = n._children - pre_start_Tasks
+        # assert len(new_tasks) == 1
+        # task = new_tasks.pop()
+
+        n: Nursery = self._n
+        cs = CancelScope()
+        new_task: Task | None = None
+        to_return: tuple[Any] | None = None
+        maybe_outcome = MaybeOutcome()
+
+        sm = self.scope_manager
+        if sm is None:
+            mngr = nullcontext([cs])
+        else:
+            mngr = sm(
+                nursery=n,
+                scope=cs,
+                maybe_outcome=maybe_outcome,
+            )
+
+        async def _start_wrapped_in_scope(
+            task_status: TaskStatus[
+                tuple[CancelScope, Task]
+            ] = trio.TASK_STATUS_IGNORED,
+
+        ) -> None:
+            nonlocal maybe_outcome
+            nonlocal to_return
+
+            with cs:
+
+                task = trio.lowlevel.current_task()
+                self._scopes[cs] = task
+
+                # TODO: instead we should probably just use
+                # `Outcome.send(mngr)` here no and inside a custom
+                # decorator `@trio.cancel_scope_manager` enforce
+                # that it's a single yield generator?
+                with mngr as to_return:
+
+                    # TODO: relay through whatever the
+                    # started task passes back via `.started()` ?
+                    # seems like that won't work with also returning
+                    # a "task handle"?
+                    task_status.started()
+
+                    # invoke underlying func now that cs is entered.
+                    outcome = await acapture(async_fn, *args)
+
+                    # TODO: instead, mngr.send(outcome) so that we don't
+                    # tie this `.start_soon()` impl to the
+                    # `MaybeOutcome` type? Use `Outcome.send(mngr)`
+                    # right?
+                    maybe_outcome._set_outcome(outcome)
+
+        await n.start(_start_wrapped_in_scope)
+        assert to_return is not None
+
+        # TODO: better way to concat the values delivered by the user
+        # provided `.scope_manager` and the outcome?
+        return tuple([maybe_outcome] + to_return)
```

Comment:
yeah as mentioned in the TODO, don't love this concat-ing of the results from the `scope_manager()`, but i think this goes away if we instead implement our own generator as `@trio.cancel_scope_manager` protocol since then the absolute value `tuple` yielded can be returned directly?